snappy/DiskCache_DiskCache.py

```diff
@@ -250,45 +252,91 @@ def getSymbolFileRelPath(self, libName, breakpadId, fileName):
-                        continue
-                    headers = response.info()
-                    contentEncoding = headers.get("Content-Encoding", "").lower()
-                    if contentEncoding in ("gzip", "x-gzip", "deflate"):
-                        with contextlib.closing(StringIO(response.read())) as dataStream:
-                            try:
-                                with gzip.GzipFile(fileobj=dataStream) as f:
-                                    data = f.read()
-                            except Exception:
-                                data = dataStream.decode('zlib')
-                    else:
-                        data = response.read()
-                libId = "{}/{}/{}".format(libName, breakpadId, symbolFilename)
-                with open(destPath, 'wb') as fp:
-                    if saveRaw:
-                        fp.write(data)
-                    else:
-                        self.writeSymMap(data.splitlines(), fp, libId)
-                self.cache.add(destPath)
-                return True
-            except Exception as e:
-                logger.log(logLevel.ERROR,
-                           "Exception when requesting symbol file at {}: {}"
-                           .format(url, e))
-                continue
-        # Ran out of URLs to try
-        return False
+        success, data = self.retrieveFile(libName, breakpadId, symbolFilename)
+        if not success:
+            return False
+
+        if not saveRaw:
+            libId = "{}/{}/{}".format(libName, breakpadId, symbolFilename)
+            data = data.splitlines()
+            data = self.makeSymMap(data, libId)
+        try:
+            destDir = os.path.dirname(destPath)
+            if not os.path.exists(destDir):
+                os.makedirs(destDir)
+            with open(destPath, 'wb') as fp:
+                fp.write(data)
+        except (OSError, IOError) as e:
+            logger.log(logLevel.ERROR, "Failed to write file {}: {}".format(destPath, e))
+            return False
+        self.cache.add(destPath)
+        return True
+
+    def retrieveFile(self, libName, breakpadId, symbolFilename):
+        """ Returns a tuple: |success, data|
+        """
+        skipURLs = []
+        for attempt in xrange(config['retries']):
+            for symbolURL in self.symbolURLs:
+                if symbolURL in skipURLs:
+                    continue
+                url = self.getSymbolURL(symbolURL, libName, breakpadId, symbolFilename)
+                success, exists, data = self.fetchURL(url)
+
+                if not success:
+                    continue
+                if not exists:
+                    # Don't retry this server if we know the file is not on it
+                    skipURLs.append(symbolURL)
+                    continue
+                return True, data
+            if config['retryDelayMs']:
+                time.sleep(config['retryDelayMs'] / 1000)
```

Comment:
I don't know enough about Tornado and how it's doing what it's doing to know whether mixing whatever it has with Python threads works well or is problematic. I skimmed the Tornado docs and my cursory glance suggests you shouldn't use Python threading, but should use the Tornado async stuff instead. It's possible that you don't see problems now because we're not running DiskCache under any real load? I'm guessing wildly.

I think i would either track down someone who knows Tornado well to help figure this out or I'd pour over the Tornado docs and rework things accordingly.